engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/consumerrecord/ConsumerRecordDeserializationSchemaFactory.scala

```diff
@@ -0,0 +1,68 @@
+package pl.touk.nussknacker.engine.kafka.consumerrecord
+
+import com.github.ghik.silencer.silent
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.streaming.connectors.kafka.KafkaDeserializationSchema
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.serialization.Deserializer
+import pl.touk.nussknacker.engine.kafka.KafkaConfig
+import pl.touk.nussknacker.engine.kafka.serialization.KafkaDeserializationSchemaFactory
+
+import scala.annotation.nowarn
+import scala.reflect.classTag
+
+/**
+  * Produces deserialization schema that describes how to turn the Kafka raw [[org.apache.kafka.clients.consumer.ConsumerRecord]]
+  * (with raw key-value of type Array[Byte]) into deserialized ConsumerRecord[K, V] (with proper key-value types).
+  * It allows the source to provide event value AND event metadata to the stream.
+  *
+  * Deprecated annotations for checksum field mapping, to fulfill constructor requirements.
+  *
+  * @tparam K - type of key of deserialized ConsumerRecord
+  * @tparam V - type of value of deserialized ConsumerRecord
+  */
+@silent("deprecated")
+@nowarn("cat=deprecation")
+abstract class ConsumerRecordDeserializationSchemaFactory[K, V] extends KafkaDeserializationSchemaFactory[ConsumerRecord[K, V]] with Serializable {
+
+  protected def createKeyDeserializer(kafkaConfig: KafkaConfig): Deserializer[K]
+
+  protected def createValueDeserializer(kafkaConfig: KafkaConfig): Deserializer[V]
+
+  override def create(topics: List[String], kafkaConfig: KafkaConfig): KafkaDeserializationSchema[ConsumerRecord[K, V]] = {
+
+    new KafkaDeserializationSchema[ConsumerRecord[K, V]] {
+
+      @transient
+      private lazy val keyDeserializer = createKeyDeserializer(kafkaConfig)
+      @transient
+      private lazy val valueDeserializer = createValueDeserializer(kafkaConfig)
+
+      override def deserialize(record: ConsumerRecord[Array[Byte], Array[Byte]]): ConsumerRecord[K, V] = {
+        val key = keyDeserializer.deserialize(record.topic(), record.key())
+        val value = valueDeserializer.deserialize(record.topic(), record.value())
+        new ConsumerRecord[K, V](
```

Comment:
Why you haven't used constructor with `leaderEpoch` and why not to also put it to `InputMeta`?

Reply:
WDYT about adding `timestampType` (with values `NO_TIMESTAMP_TYPE(-1, "NoTimestampType"), CREATE_TIME(0, "CreateTime"), LOG_APPEND_TIME(1, "LogAppendTime")`)?

Reply:
IMO more information is always better option then less in case of meta data :+1: